### DISCUSSION TITLE
How is "{{- index .Annotations \"otel-service-name\" -}}" resolved

### DISCUSSION THREAD

**Discussion by paulomorgado** [microsoft/aspire#2046] [Q&A] [answered]
When adding a container, to add Otlp exporter, `WithOtlpExporter()` needs to be invoked on the container builder.

`WithOtlpExporter()`, in turn, invokes `AddOtlpEnvironment` that ends up adding this environment variable:

```csharp
context.EnvironmentVariables["OTEL_SERVICE_NAME"] = "{{- index .Annotations \"otel-service-name\" -}}";
```

When and how is `"{{- index .Annotations \"otel-service-name\" -}}"`? Because it's empty for containers.

My workaround was to invoke `WithEnvironment("OTEL_SERVICE_NAME", "service name")`, but it feels odd.

**Comment by davidfowl** [comment]
cc @karolz-ms

**Comment by karolz-ms** [comment]
`otel-service-name` annotation, which is the source for the `OTEL_SERVICE_NAME` environment variable is applied when the Aspire model is translated into DCP model for local execution. Specifically [here](https://github.com/dotnet/aspire/blob/main/src/Aspire.Hosting/Dcp/ApplicationExecutor.cs#L494) and [here]( https://github.com/dotnet/aspire/blob/main/src/Aspire.Hosting/Dcp/ApplicationExecutor.cs#L466). The motivation behind is, when a service has multiple replicas, all replicas should use the same OTEL service name for telemetry--they are instances of the same thing from telemetry perspective. So we add this annotation to the "replica set" object, which then gets propagated to all replicas, and this is how replicas end up knowing which "service name" they should be using when reporting telemetry. 

That said, in the current code the annotation is only added to objects that get run as ordinary processes, not containers. That's probably a bug and we should probably be adding that annotation to everything that might export telemetry. @jamesnk @mitchdenny  what do you guys think?

**Reply by paulomorgado** [reply]
These are private methods the user has no access to.

The annotation is added for container resources, but there doesn't seem to be a way to resolve it. The user needs to explicitly add the environment variable.

**Reply by karolz-ms** [reply, marked as answer]
Right, I think the workaround you have is the best solution for now. Sorry, I should have been more specific.

**Reply by davidfowl** [reply]
We should fix this.
